consensus/src/quorum_store/batch_generator.rs

```diff
@@ -95,18 +102,98 @@ impl BatchGenerator {
         }
     }
 
-    pub(crate) async fn handle_scheduled_pull(&mut self, max_count: u64) -> Option<Batch> {
-        // TODO: as an optimization, we could filter out the txns that have expired
+    fn create_new_batch(
+        &mut self,
+        txns: Vec<SignedTransaction>,
+        expiry_time: u64,
+        bucket_start: u64,
+    ) -> Batch {
+        let batch_id = self.batch_id;
+        self.batch_id.increment();
+        self.db
+            .save_batch_id(self.epoch, self.batch_id)
+            .expect("Could not save to db");
+
+        let txns_in_progress: Vec<_> = txns
+            .iter()
+            .map(|txn| TransactionInProgress {
+                summary: TransactionSummary {
+                    sender: txn.sender(),
+                    sequence_number: txn.sequence_number(),
+                },
+                gas_unit_price: txn.gas_unit_price(),
+            })
+            .collect();
+        self.batches_in_progress.insert(batch_id, txns_in_progress);
+        self.batch_expirations.add_item(batch_id, expiry_time);
+
+        counters::CREATED_BATCHES_COUNT.inc();
+        counters::num_txn_per_batch(bucket_start.to_string().as_str(), txns.len());
+
+        Batch::new(
+            batch_id,
+            txns,
+            self.epoch,
+            expiry_time,
+            self.my_peer_id,
+            bucket_start,
+        )
+    }
+
+    /// Push num_txns from txns into batches. If num_txns is larger than max size, then multiple
+    /// batches are pushed.
+    fn push_to_batches(
+        &mut self,
+        batches: &mut Vec<Batch>,
+        txns: &mut Vec<SignedTransaction>,
+        num_txns: usize,
+        expiry_time: u64,
+        bucket_start: u64,
+    ) {
+        let mut remaining_txns = num_txns;
+        while remaining_txns > 0 {
+            let num_batch_txns = std::cmp::min(self.config.max_batch_txns, remaining_txns);
+            let batch_txns: Vec<_> = txns.drain(0..num_batch_txns).collect();
+            let batch = self.create_new_batch(batch_txns, expiry_time, bucket_start);
+            batches.push(batch);
+            remaining_txns -= num_batch_txns;
+        }
+    }
+
+    fn build_account_min_gas(
+        exclude_txns: &[TransactionInProgress],
```

Comment:
Some txns here might be already committed. So min might be wrong due to race. Is that an issue?